sys/posix: add module posix_netdb

HendrikVE opened the pull request:
### Contribution description

This PR adds a new posix module which provides netdb functions like `gethostbyname` or `getaddrinfo`. It does not come with its own implementation. It is a meta module collecting implementations, at the moment only form LwIP. I decided to do it this way so the code does not have to be copy&pasted over into the RIOT repository. Doing so would lead to the problem that the module would be getting out of sync with the LwIP code once its updated.
I also added the option to use posix sockets for the existing `sock_dns` module. Without using posix sockets I had extremely poor success rates for address resolution (approximately 1 out of 20).

### Testing procedure

You can test this PR e.g. in conjunction with everything else in #15969.

### Issues/PRs references

Missing posix feature, see #7801.
This PR is split out from the chunky #15969.

miri64 reviewed (changes requested):
> I also added the option to use posix sockets for the existing `sock_dns` module. Without using posix sockets I had extremely poor success rates for address resolution (approximately 1 out of 20).

NACK. 1. We rather should find the reason, why `sock_udp` performs so poorly compared to POSIX sockets (which below use `sock_udp`, so this fact seems weird to me). 2. `sock_dns` using POSIX sockets turns everything on its head, now we have a `sock` module using POSIX sockets using a `sock` module. Conceptually alone this sounds very dirty.

HendrikVE commented:
I have to correct myself. Either I did not remember correctly (has been some months now) or it is due to the rebase, but without using posix sockets it is not working at all. I traced it back to the function `sys_arch_mbox_fetch` at `pkg/lwip/contrib/sys_arch.c` which is repeated a few times and then gets stuck for some reason. I agree it is dirty, I should have mentioned that this is one of those hacks that should be removed in the end ;) But I needed this for now in order to proceed with the wolfmqtt port.
I pushed a throwaway commit to #15969. In the file `dns.c` you will find the line `#define USE_POSIX 0` where you can switch between the original module and the posix extension. Any idea what could cause this?

miri64 commented:
Ok, so the problem is with lwIP's `sock` implementation. That at least tells me, that `sock` is not completely broken ;-). As to `sys_arch_mbox_fetch`: It is very similar to what `gnrc_sock_recv()` does in `gnrc_sock.c` (the first is just more generalized). Could you check if there is any significant difference (e.g. #16831 was merged recently there) that could make a difference in your case? Maybe it is about time we add a `xtimer`/`ztimer` function for that kind of functionality (`xtimer_mbox_get_timeout()`...).

miri64 commented:
> In the file `dns.c` you will find the line `#define USE_POSIX 1` where you can switch between the original module and the posix extension.

Still a dirty hack that should not go into `sock_dns`. Then I would rather have a new `posix_dns` module. Due to the DNS message being factored out to `dns_msg` now, that should be easy enough :-). But I guess in that case it makes more sense to use implement it in `posix_netdb` directly.

miri64 commented:
> Due to the DNS message being factored out to `dns_msg` now, that should be easy enough :-).

Speaking of, I really think you should rebase this PR ;-).

HendrikVE commented:
> Ok, so the problem is with lwIP's `sock` implementation. That at least tells me, that `sock` is not completely broken ;-). As to `sys_arch_mbox_fetch`: It is very similar to what `gnrc_sock_recv()` does in `gnrc_sock.c` (the first is just more generalized). Could you check if there is any significant difference (e.g. #16831 was merged recently there) that could make a difference in your case? Maybe it is about time we add a `xtimer`/`ztimer` function for that kind of functionality (`xtimer_mbox_get_timeout()`...).

I'll have a look into it

HendrikVE commented:
> > In the file `dns.c` you will find the line `#define USE_POSIX 1` where you can switch between the original module and the posix extension.
> 
> Still a dirty hack that should not go into `sock_dns`. Then I would rather have a new `posix_dns` module. Due to the DNS message being factored out to `dns_msg` now, that should be easy enough :-). But I guess in that case it makes more sense to use implement it in `posix_netdb` directly.

This was just so one could test this with a working wolfmqtt application ;) I would rather find out why this is happening, but if I can't find out the reason in reasonable time and if you say it is a viable option I might resort to implement a `posix_dns` module.

miri64 commented:
> This was just so one could test this with a working wolfmqtt application ;) I would rather find out why this is happening, but if I can't find out the reason in reasonable time and if you say it is a viable option I might resort to implement a `posix_dns` module.

Again, I think the more viable option in the latter case would be, I think to put the functionality directly into `posix_netdb` /  `gethostbyname` instead of adding another module that would implement non-standard functions.

miri64 commented on pkg/lwip/patches/0002-adapt-netdb-functions-for-RIOT.patch:
```diff
@@ -0,0 +1,249 @@
+   HOSTENT_STORAGE char s_hostname[DNS_MAX_NAME_LENGTH + 1];
+ 
+   /* query host IP address */
+-  err = netconn_gethostbyname(name, &addr);
+-  if (err != ERR_OK) {
++  int res = sock_dns_query(name, &addr, AF_INET);
++  if (res < 0) {
```
I don't quite understand this: why can't `netconn_gethostbyname()` be used here?

miri64 commented on sys/posix/netdb/netdb.c:
```diff
@@ -0,0 +1,55 @@
+#include "netdb.h"
+
+struct hostent* gethostbyname(const char *name) {
+#if LWIP_NETDB
+    return lwip_gethostbyname(name);
+#else
+    #error "There is not implementation of gethostbyname available"
+#endif
```
For all the functions here: I think it might be cleaner (and smaller) to provide our own implementation utilizing either `sock_dns` or `sock_udp` directly here. This would have multiple advantages:

- no stack-wise `#ifdefs`
- other DNS protocols than the standard DNS over UDP (see #16705 and #16861) and even completely different name-resolution protocols ([`netdb.h` spec does not mention DNS at all](https://pubs.opengroup.org/onlinepubs/9699919799.2018edition/basedefs/netdb.h.html) and [`getaddrinfo` spec just says "In many cases it is implemented by the Domain Name System, […]"](https://pubs.opengroup.org/onlinepubs/9699919799.2018edition/functions/getaddrinfo.html)) could be used here.

BTW, since I looked at the specs: `gethostbyname()` and its re-entrant counter-part where removed in POSIX.1-2008, so do we really want to support them and not rather fix libraries that still use them?

HendrikVE commented:
Thank you for your suggestions.
Took me some time to figure out the problem with the program getting stuck on `sys_arch_mbox_fetch`. That's what I thought at least... In fact `sys_arch_mbox_fetch` does return, but not back to `netconn_recv_data` were it was originally called from. Instead `netconn_new_with_proto_and_callback` is magically called in between. Seems like lwIP (or we?) does some dark magic with return pointers. Anyway, this lead me to the root of the problem.

The cause of this problem was the fact that I am using `posix_select` for my wolfmqtt port. `USEMODULE += posix_select` leads to `USEMODULE += sock_async`, which leads to `USEMODULE += lwip_sock_async` which will define `CFLAGS += -DSOCK_HAS_ASYNC`. So, turns out `sock->base.async_cb.gen` wasn't initialized to NULL. Because of that it had some garbage value != NULL and was called by `_netconn_cb` and lead to the program being stopped.

Anyway, I dropped `gethostbyname` as you suggested and we now have our own implementation of `getaddrinfo`, which is a combination of lwIP's and glibc's implementation. I also added a test for the module.

OlegHahm commented:
@miri64, what's your latest take on this one?

HendrikVE commented:
The last open todo is now finished. In exchange, the `posix_netdb` module now needs a few prep commits. If @miri64 prefers I could also split them into separate PRs.

All green btw ;)

HendrikVE commented:
@benpicco For the case that this PR is not further split: I turned  your dns mock code into a pseudomodule `sock_dns_mock`, so that `tests/posix_netdb` can make use of it as well. You might want to have a look at this.

benpicco reviewed (commented):
Maybe the description needs an update, but what's LWIP specific about this?

benpicco commented on sys/posix/netdb/netdb.c:
```diff
@@ -0,0 +1,277 @@
+        return EAI_BADFLAGS;
+    }
+
+    if ((   hints->ai_family != AF_UNSPEC)
+        && (hints->ai_family != AF_INET)
+        && (hints->ai_family != AF_INET6)
+) {
```
```suggestion
    if ((hints->ai_family != AF_UNSPEC) &&
        (hints->ai_family != AF_INET) &&
        (hints->ai_family != AF_INET6)) {
```

benpicco commented on tests/posix_netdb/Makefile.board.dep:
```diff
@@ -0,0 +1,9 @@
+endif
+
+ifeq ($(BOARD),native)
+  USEMODULE += netdev_default
```
Why only for `native`?

benpicco commented:
CI is unhappy

HendrikVE commented:
> CI is unhappy

Seems like the esp32 platform, in contrast to the other platforms, defines its own `errno.h` in `../../../build/pkg/esp32_sdk/components/newlib/platform_include/errno.h`. We still need do provide these values for all the other platforms though. So either we check for every value whether it is set already or we think about a way how to handle `errno.h` in general, like providing a RIOT-global `errno.h`. I'd prefer the latter.

benpicco commented on sys/posix/include/netdb.h:
```diff
@@ -0,0 +1,179 @@
+/**
+ * @brief The name could not be resolved at this time. Future attempts may succeed.
+ */
+#define EAI_AGAIN       1
```
```suggestion
#ifndef EAI_AGAIN 
#define EAI_AGAIN       1
#endif
```

would be the simplest solution

benpicco reviewed (commented):
I'm not sure what you mean by providing `errno.h` in RIOT - it's already provided by libc

HendrikVE commented:
> I'm not sure what you mean by providing `errno.h` in RIOT - it's already provided by libc

If [this source](https://chromium.googlesource.com/native_client/nacl-newlib/+/master/newlib/libc/include/sys/errno.h) is correct, then libc defines `EAGAIN` but not `EAI_AGAIN`. Futhermore if one looks for it, `EAI_AGAIN` is directly associated with `getaddrinfo`/`posix`. It's strange that the esp32 would include these posix errors in its `errno.h` then, isn't it?

HendrikVE commented:
For some weird reason the three error codes `EAI_SOCKTYPE`, `EAI_AGAIN` and `EAI_BADFLAGS` are declared in esp32's [`errno.h`](https://github.com/espressif/esp-idf/blob/master/components/newlib/platform_include/errno.h). `ifndef` would not fix this issue, since I would need to adapt to the same values as given by the esp32. Since esp32 defines `EAI_BADFLAGS=3` it would have the same value as my `EAI_FAIL` which is really bad.

@gschorcht Can we get rid of these error codes for the esp32?

miri64 commented on sys/posix/netdb/netdb.c:
```diff
@@ -0,0 +1,276 @@
+            }
+        }
+        else { /* do actual DNS lookup */
+            uint8_t addr_buf[16] = { 0 };
+            int res = dns_query(nodename, addr_buf, hints->ai_family);
+            switch (res) {
+            case -ETIMEDOUT:
+                /* fall-through */
+            case -EAGAIN:
+                return EAI_AGAIN;
+
+            case -ENOMEM:
+                return EAI_MEMORY;
+
+            case INADDRSZ:
+                addr.type = AF_INET;
+                memcpy(&addr.ipv4, addr_buf, INADDRSZ);
+                break;
+
+            case IN6ADDRSZ:
+                addr.type = AF_INET6;
+                memcpy(&addr.ipv6, addr_buf, IN6ADDRSZ);
+                break;
```
Why not use `addr` directly?

Make `ipvx_addr_t` the following:

```C
typedef struct ipvx_addr {
    sa_family_t type;
    union {
        struct in_addr ipv4;
        struct in6_addr ipv6;
        uint8_t u8[16];
    };
} ipvx_addr_t;
```
Then just do
```suggestion
            int res = dns_query(nodename, addr->u8, hints->ai_family);
            switch (res) {
            case -ETIMEDOUT:
                /* fall-through */
            case -EAGAIN:
                return EAI_AGAIN;
            case -ENOMEM:
                return EAI_MEMORY;
            case INADDRSZ:
                addr.type = AF_INET;
                break;
            case IN6ADDRSZ:
                addr.type = AF_INET6;
                break;
```

And safe both a little ROM and RAM

miri64 commented on sys/posix/netdb/netdb.c:
```diff
@@ -0,0 +1,276 @@
+    if (nodename != NULL) {
+        name_len = strlen(nodename);
+        if (name_len > SOCK_DNS_MAX_NAME_LEN) {
+            /* invalid name length */
+            return EAI_FAIL;
```
Maybe even use an error goto?

At the very end of your function put

```
error:
    return errno;
```

And then, whenever there is an error (such in this case) use

```suggestion
            /* invalid name length */
            errno = EAI_FAIL;
            goto error;
```

miri64 commented on sys/posix/netdb/netdb.c:
```diff
@@ -0,0 +1,276 @@
+        return EAI_MEMORY;
+    }
+
+    if (addr.type == AF_INET) {
+        ai->ai_family = AF_INET;
+
+        struct sockaddr_in *sa4 = (struct sockaddr_in*) sas;
+        sa4->sin_addr = addr.ipv4;
+        sa4->sin_family = AF_INET;
+        sa4->sin_port = htons(port);
+    }
+    else if (addr.type == AF_INET6) {
+        ai->ai_family = AF_INET6;
+
+        struct sockaddr_in6 *sa6 = (struct sockaddr_in6*) sas;
+        sa6->sin6_addr = addr.ipv6;
+        sa6->sin6_family = AF_INET6;
+        sa6->sin6_port = htons(port);
+    }
```
```suggestion
    ai->ai_family = addr.type;

    struct sockaddr_in6 *sa = (struct sockaddr_in6*) sas;
    sa->sin6_addr = addr.ipv6;
    sa->sin6_family = addr.type;
    sa->sin6_port = htons(port);
```

Should suffice.

miri64 commented on sys/posix/netdb/netdb.c:
```diff
@@ -0,0 +1,276 @@
+            return EAI_MEMORY;
+        }
+
+        memcpy(ai->ai_canonname, nodename, name_len);
```
Nitpick: Use `strncpy()` instead? You are copying a string here after all.

miri64 commented on tests/posix_netdb/Makefile:
```diff
@@ -0,0 +1,17 @@
+USEMODULE += embunit
+
+# make sure we have an implementation of sock_types.h
+USEMODULE += sock_udp
+USEMODULE += gnrc_ipv6
+
```
Do you need the whole stack if you mock DNS anyway?

miri64 commented on tests/posix_netdb/main.c:
```diff
@@ -0,0 +1,294 @@
+    TEST_ASSERT_EQUAL_INT(AF_INET, result->ai_family);
+    TEST_ASSERT_NOT_NULL(inet_ntop(AF_INET, &((struct sockaddr_in*)(result->ai_addr))->sin_addr,
+                                   addr4_str, sizeof(addr4_str)));
+    TEST_ASSERT_EQUAL_INT(0, strcmp(addr4_str, test_ipv4));
```
Here and in other instances where you used `TEST_ASSERT_EQUAL_INT(0, strcmp(...))`

```suggestion
    TEST_ASSERT_EQUAL_STRING(addr4_str, test_ipv4);
```

miri64 commented on tests/posix_netdb/main.c:
```diff
@@ -0,0 +1,294 @@
+    TEST_ASSERT_EQUAL_INT(0, getaddrinfo(test_ipv6, NULL, &hints, &result));
+    TEST_ASSERT_NOT_NULL(result);
+    TEST_ASSERT_EQUAL_INT(AF_INET6, result->ai_family);
+    TEST_ASSERT_NOT_NULL(inet_ntop(AF_INET6, &((struct sockaddr_in6*)(result->ai_addr))->sin6_addr,
+                                   addr6_str, sizeof(addr6_str)));
```
Maybe use a temporary variable here for sockaddr_in6 for better readability?

miri64 commented on tests/posix_netdb/main.c:
```diff
@@ -0,0 +1,294 @@
+
+    switch (result->ai_family) {
+    case AF_INET:
+        memcpy(&addr_ipv4, &((struct sockaddr_in*)(result->ai_addr))->sin_addr.s_addr, sizeof(ipv4_addr_t));
```
Dito (also: in its current state the line is longer than 100 chars)

miri64 commented on tests/posix_netdb/main.c:
```diff
@@ -0,0 +1,294 @@
+                                   addr4_str, sizeof(addr4_str)));
+    TEST_ASSERT_EQUAL_INT(0, strcmp(addr4_str, test_ipv4));
+    freeaddrinfo(result);
+
+
```
```suggestion

```

miri64 commented on tests/posix_netdb/Makefile:
```diff
@@ -0,0 +1,17 @@
+include ../Makefile.tests_common
+
+USEMODULE += embunit
+
+# make sure we have an implementation of sock_types.h
+USEMODULE += sock_udp
+USEMODULE += gnrc_ipv6
+
+USEMODULE += ipv4_addr
+USEMODULE += ipv6_addr
+
+# use mock data for the test
+USEMODULE += sock_dns_mock
+
+USEMODULE += posix_netdb
+
+include $(RIOTBASE)/Makefile.include
```
```bash
make -C tests/posix_netdb/ all-valgrind term-valgrind
```

yields

```
--105511-- warning: evaluate_Dwarf3_Expr: unhandled DW_OP_ 0x9b
--105511-- warning: evaluate_Dwarf3_Expr: unhandled DW_OP_ 0x9b
--105511-- warning: evaluate_Dwarf3_Expr: unhandled DW_OP_ 0x9b
--105511-- warning: evaluate_Dwarf3_Expr: unhandled DW_OP_ 0x9b
--105511-- warning: evaluate_Dwarf3_Expr: unhandled DW_OP_ 0x9b
--105511-- warning: evaluate_Dwarf3_Expr: unhandled DW_OP_ 0x9b
--105511-- warning: evaluate_Dwarf3_Expr: unhandled DW_OP_ 0x9b
--105511-- warning: evaluate_Dwarf3_Expr: unhandled DW_OP_ 0x9b
--105511-- warning: evaluate_Dwarf3_Expr: unhandled DW_OP_ 0x9b
--105511-- warning: evaluate_Dwarf3_Expr: unhandled DW_OP_ 0x9b
--105511-- warning: evaluate_Dwarf3_Expr: unhandled DW_OP_ 0x9b
==105511== Invalid read of size 4
==105511==    at 0x804BC0C: ??? (in /home/mlenders/Repositories/RIOT-OS/RIOT/tests/posix_netdb/bin/native/tests_posix_netdb.elf)
==105511==  Address 0x806c790 is in the brk data segment 0x8079000-0x8078fff
==105511== 
{ "threads": [{ "name": "main", "stack_size": 12288, "stack_used": 2444}]}
```

Sadly no line number but somewhere is an out of bounds read error to one of the `malloc`'d memory areas in this test (or the netdb code).
